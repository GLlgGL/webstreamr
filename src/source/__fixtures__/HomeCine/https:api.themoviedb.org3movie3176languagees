{"adult":false,"backdrop_path":"/amBvmIshdsSkOtvVIgxl7YSQ9Dg.jpg","belongs_to_collection":{"id":16302,"name":"Battle Royale - Colección","poster_path":"/e9x8MueDBGtEAALdZZIZQ96i8C6.jpg","backdrop_path":"/gKpdj5aKtxrAoAbJJ3EtBZHcsyP.jpg"},"budget":4500000,"genres":[{"id":18,"name":"Drama"},{"id":53,"name":"Suspense"},{"id":28,"name":"Acción"}],"homepage":"","id":3176,"imdb_id":"tt0266308","origin_country":["JP"],"original_language":"ja","original_title":"バトル・ロワイアル","overview":"En el amanecer de un nuevo milenio, el país está al borde del colapso. Millones de personas vagan sin empleo. La violencia en la escuela está descontrolada y adolescentes rebeldes protagonizan boicots masivos. El gobierno contrataca con “Battle Royale”. Cada año, una clase es escogida al azar para que se enfrente, en una isla abandonada, a un cruel juego de supervivencia.","popularity":7.5811,"poster_path":"/e9UQGoX2ZFNdrjLwEejMSMDSxmR.jpg","production_companies":[{"id":5822,"logo_path":"/qyTbRgCyU9NLKvKaiQVbadtr7RY.png","name":"Toei Company","origin_country":"JP"},{"id":157807,"logo_path":"/tyrJOOVAUIxQEJzjP1PhP2H5zFi.png","name":"WOWOW","origin_country":"JP"},{"id":11538,"logo_path":null,"name":"AM Associates","origin_country":"JP"},{"id":11539,"logo_path":null,"name":"Kobi","origin_country":"JP"},{"id":3032,"logo_path":"/gaf3ez34cuvG9snphWGreB1WerA.png","name":"Nippan Group Holdings","origin_country":"JP"},{"id":11540,"logo_path":null,"name":"MF Pictures","origin_country":"JP"},{"id":3656,"logo_path":"/mY6L0bHRp8fNWfICPqLjxp3khvq.png","name":"GAGA Communications","origin_country":"JP"}],"production_countries":[{"iso_3166_1":"JP","name":"Japan"}],"release_date":"2000-12-16","revenue":30600000,"runtime":114,"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"},{"english_name":"Japanese","iso_639_1":"ja","name":"日本語"}],"status":"Released","tagline":"¿Puedes matar a tu mejor amigo?","title":"Battle Royale","video":false,"vote_average":7.3,"vote_count":3603}