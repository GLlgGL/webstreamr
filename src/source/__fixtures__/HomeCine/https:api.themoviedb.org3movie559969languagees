{"adult":false,"backdrop_path":"/uLXK1LQM28XovWHPao3ViTeggXA.jpg","belongs_to_collection":null,"budget":0,"genres":[{"id":80,"name":"Crimen"},{"id":18,"name":"Drama"},{"id":53,"name":"Suspense"}],"homepage":"https://www.netflix.com/es/title/81078819","id":559969,"imdb_id":"tt9243946","origin_country":["US"],"original_language":"en","original_title":"El Camino: A Breaking Bad Movie","overview":"Tiempo después de los eventos sucedidos tras el último episodio de la serie \"Breaking Bad\", el fugitivo Jesse Pinkman (Aaron Paul) huye de sus perseguidores, de la ley y de su pasado.","popularity":9.9122,"poster_path":"/hoWADuvXs3Ua4AXBAiZYnppTupO.jpg","production_companies":[{"id":11073,"logo_path":"/wHs44fktdoj6c378ZbSWfzKsM2Z.png","name":"Sony Pictures Television","origin_country":"US"},{"id":2605,"logo_path":null,"name":"Gran Via Productions","origin_country":"US"},{"id":33742,"logo_path":"/2jdh2sEa0R6y6uT0F7g0IgA2WO8.png","name":"High Bridge Productions","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"release_date":"2019-10-11","revenue":0,"runtime":123,"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Released","tagline":"","title":"El Camino: Una película de Breaking Bad","video":false,"vote_average":7.0,"vote_count":5192}