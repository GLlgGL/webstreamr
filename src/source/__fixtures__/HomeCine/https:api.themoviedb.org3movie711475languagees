{"adult":false,"backdrop_path":"/a8Bbbpqi6sd58Nr4BzbMQa8VFPp.jpg","belongs_to_collection":null,"budget":0,"genres":[{"id":35,"name":"Comedia"},{"id":18,"name":"Drama"}],"homepage":"","id":711475,"imdb_id":"tt10944596","origin_country":["US"],"original_language":"en","original_title":"Here Today","overview":"El veterano escritor de comedia Charlie Berns, que poco a poco está perdiendo el control de la realidad, se hace amigo de una joven y talentosa cantante callejera de Nueva York, Emma Payge. Juntos, forman una amistad poco probable pero divertida y conmovedora que hace a un lado la brecha generacional y redefine el significado del amor y la confianza.","popularity":1.6152,"poster_path":"/7kOjuw40icNKtkviDj7WkPaOb2.jpg","production_companies":[{"id":106405,"logo_path":"/loAbuz9SPCa9rrDPskWBOQ6rJlI.png","name":"Astute Films","origin_country":"US"},{"id":59935,"logo_path":null,"name":"Big Head Productions","origin_country":""},{"id":11341,"logo_path":"/xytTBODEy3p20ksHL4Ftxr483Iv.png","name":"Stage 6 Films","origin_country":"US"},{"id":48638,"logo_path":null,"name":"Face Productions","origin_country":"US"},{"id":6831,"logo_path":"/fWSI4EjIDEEzCNJC8OTPX9Pdbxv.png","name":"Big Indie Pictures","origin_country":"US"},{"id":52059,"logo_path":"/rajoRd4i5xJKHAqFyeflFBMUBZM.png","name":"RocketScience","origin_country":"GB"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"release_date":"2021-05-07","revenue":0,"runtime":117,"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Released","tagline":"Nunca se sabe quién va a cambiar tu vida.","title":"Aquí y ahora","video":false,"vote_average":7.0,"vote_count":153}