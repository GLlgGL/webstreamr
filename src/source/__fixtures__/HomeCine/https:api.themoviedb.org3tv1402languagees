{"adult":false,"backdrop_path":"/rAOjnEFTuNysY7bot8zonhImGMh.jpg","created_by":[{"id":4027,"credit_id":"5256cb3f19c2956ff605e5de","name":"Frank Darabont","original_name":"Frank Darabont","gender":2,"profile_path":"/vZ50guP86otYTiBSGfi35GNHWVf.jpg"}],"episode_run_time":[],"first_air_date":"2010-10-31","genres":[{"id":10759,"name":"Action & Adventure"},{"id":18,"name":"Drama"},{"id":10765,"name":"Sci-Fi & Fantasy"}],"homepage":"http://www.amc.com/shows/the-walking-dead--1002293","id":1402,"in_production":false,"languages":["en"],"last_air_date":"2022-11-20","last_episode_to_air":{"id":3979137,"name":"Descansa en paz","overview":"Daryl y Carol llevan a Judith al hospital; Rosita, Eugene y Gabriel buscan a Coco; Maggie y Negan toman las armas contra Pamela; los héroes se reúnen para una última batalla.","vote_average":8.6,"vote_count":85,"air_date":"2022-11-20","episode_number":24,"episode_type":"finale","production_code":"","runtime":65,"season_number":11,"show_id":1402,"still_path":"/8BUXYeIeRajrgfJawYXCPAMwzkw.jpg"},"name":"The Walking Dead","next_episode_to_air":null,"networks":[{"id":174,"logo_path":"/pmvRmATOCaDykE6JrVoeYxlFHw3.png","name":"AMC","origin_country":"US"}],"number_of_episodes":177,"number_of_seasons":11,"origin_country":["US"],"original_language":"en","original_name":"The Walking Dead","overview":"\"The Walking Dead\" está ambientada en un futuro apocalíptico con la Tierra devastada por el efecto de un cataclismo, que ha provocado la mutación en zombies de la mayor parte de los habitantes del planeta. La serie, explora las dificultades de los protagonistas para sobrevivir en un mundo poblado por el horror, así como las relaciones personales que se establecen entre ellos, en ocasiones también una amenaza para su supervivencia.","popularity":70.8638,"poster_path":"/hUblG1KZCTRpHc3wqqoU0DW98Q3.jpg","production_companies":[{"id":23242,"logo_path":"/ucE6dN6pUPfZHH5JlXkq44TDUMS.png","name":"AMC Studios","origin_country":"US"},{"id":23921,"logo_path":"/simDvqT8y6jhP530ggUMbikvVKc.png","name":"Circle of Confusion","origin_country":"US"},{"id":11533,"logo_path":"/tWM9pmzVYxok4GbQIttxdcml1yT.png","name":"Valhalla Motion Pictures","origin_country":"US"},{"id":3982,"logo_path":"/bli7HkPOXOEWsDwDK0W7XXfeUU2.png","name":"Darkwoods Productions","origin_country":"US"},{"id":50032,"logo_path":"/ojrMksdtR7Bs2vwERU1LCrjv7vx.png","name":"Skybound Entertainment","origin_country":"US"},{"id":170794,"logo_path":null,"name":"Idiotbox","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"seasons":[{"air_date":"2010-10-11","episode_count":82,"id":3646,"name":"Especiales","overview":"","poster_path":"/yNQlx9uKgqGjH1g1lP2YOMPAwfC.jpg","season_number":0,"vote_average":0.0},{"air_date":"2010-10-31","episode_count":6,"id":3643,"name":"Temporada 1","overview":"En un mundo devastado por un apocalipsis zombi, un variopinto grupo de supervivientes se unen para permanecer a salvo y obtener respuestas dentro de los muros del CCE.","poster_path":"/5eL2Bym6NZdTLs4unoinAQZsDbB.jpg","season_number":1,"vote_average":8.4},{"air_date":"2011-10-16","episode_count":13,"id":3644,"name":"Temporada 2","overview":"Durante la segunda temporada de esta lúgubre serie de zombis, Grimes y sus cohortes continúan la búsqueda del santuario en un paraje desolado y peligroso.","poster_path":"/je7QVgGENEZSW2V16sHlBWW0Ra8.jpg","season_number":2,"vote_average":8.5},{"air_date":"2012-10-13","episode_count":16,"id":3645,"name":"Temporada 3","overview":"En la temporada 3, el grupo de supervivientes de Rick toma el control de una prisión abandonada y descubre un asentamiento dirigido por el poderoso Gobernador.","poster_path":"/9AJw8IpFAVGjPCZ16hYG1sTRBMd.jpg","season_number":3,"vote_average":8.6},{"air_date":"2013-10-13","episode_count":16,"id":3647,"name":"Temporada 4","overview":"Después de la masacre, comienza la caza al Gobernador. Mientras, Rick, Daryl, Michonne y Karen planean su siguiente movimiento.","poster_path":"/s03WseSKRuOjYx8ZKbFJ5Z1dt2E.jpg","season_number":4,"vote_average":8.5},{"air_date":"2014-10-12","episode_count":16,"id":60391,"name":"Temporada 5","overview":"En la 5.ª temporada, Rick y los demás se enfrentan a sus captores en La Terminal, descubren qué pasó con Beth y deciden si confiar en nuevos grupos de supervivientes.","poster_path":"/yNNa5AnguZRBmKBZlxA38WoiDqP.jpg","season_number":5,"vote_average":8.5},{"air_date":"2015-10-11","episode_count":16,"id":68814,"name":"Temporada 6","overview":"Tras una muerte traumática, el grupo de Rick y los otros residentes de Alexandria cuestionan sus posibilidades al tiempo que procuran nuevas formas de sobrevivir.","poster_path":"/teqkAQmxqAZMNQ6NOpo9pZMQWew.jpg","season_number":6,"vote_average":8.4},{"air_date":"2016-10-22","episode_count":16,"id":76834,"name":"Temporada 7","overview":"El grupo de Rick se recupera de un acto de violencia brutal y se pregunta qué otros horrores les tienen reservados Negan y los Salvadores.","poster_path":"/q17oXRgzW739jFqCd2XBSFHqDkv.jpg","season_number":7,"vote_average":7.9},{"air_date":"2017-10-21","episode_count":16,"id":91735,"name":"Temporada 8","overview":"Mientras Negan continúa con su tiránico reinado de terror, Rick reúne a sus aliados del Reino y de Hilltop para lanzar un ataque frontal contra los Salvadores.","poster_path":"/djgM6a4t30GyWCn8SufCCxxDNlg.jpg","season_number":8,"vote_average":7.9},{"air_date":"2018-10-07","episode_count":16,"id":109531,"name":"Temporada 9","overview":"Tras vencer a los Salvadores, la decisión de dejar a Negan con vida provoca una brecha en el grupo de supervivientes. El desencuentro es evidente entre dos bandos diferenciados, liderados por Rick y Maggie respectivamente, que están obligados a convivir juntos.","poster_path":"/ir3HYeK8rdcZlndnnfouQubHG75.jpg","season_number":9,"vote_average":8.3},{"air_date":"2019-10-06","episode_count":22,"id":123967,"name":"Temporada 10","overview":"Es primavera, y han pasado unos meses desde del final de la temporada 9, cuando nuestro grupo de supervivientes se atrevió a cruzar al territorio de los Susurradores durante el duro invierno. Las comunidades aún están lidiando con las secuelas tras el temible despliegue de poder de los Alpha, y la idea misma de si la civilización podrá sobrevivir en un mundo lleno de muertos está en juego.","poster_path":"/jKVxjqXQstVLetQOTsJ6sbj8c6d.jpg","season_number":10,"vote_average":8.2},{"air_date":"2021-08-22","episode_count":24,"id":189337,"name":"Temporada 11","overview":"Alejandría está severamente comprometida tras la devastación dejada por los Susurradores. Ahora todos los que viven en Alejandría luchan por refortificarla y alimentar a su creciente número de residentes, que incluyen a los sobrevivientes de la caída del Reino y la quema de Hilltop; junto con Maggie y su nuevo grupo, los Guardianes. Alejandría tiene más personas de las que puede lograr alimentar y proteger. Su situación es grave a medida que las tensiones se calientan por los acontecimientos pasados y la autopreservación se eleva a la superficie dentro de los muros devastados.","poster_path":"/3xWqI8U11wwnC0ApqQI5qFLAvg.jpg","season_number":11,"vote_average":8.1}],"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Ended","tagline":"Lucha contra los muertos. Teme a los vivos.","type":"Scripted","vote_average":8.092,"vote_count":17207}