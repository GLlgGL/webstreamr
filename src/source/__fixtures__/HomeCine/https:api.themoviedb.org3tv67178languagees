{"adult":false,"backdrop_path":"/jBGjbSDRxOEudW9rmQbWDzJUKq9.jpg","created_by":[{"id":1240816,"credit_id":"580d4a20c3a3682286013d6b","name":"Steve Lightfoot","original_name":"Steve Lightfoot","gender":2,"profile_path":"/2CPBXSicKlQO0xnhmYgEhWUGL42.jpg"}],"episode_run_time":[53],"first_air_date":"2017-11-17","genres":[{"id":10759,"name":"Action & Adventure"},{"id":80,"name":"Crimen"},{"id":18,"name":"Drama"}],"homepage":"https://www.disneyplus.com/series/the-punisher/6JOZHgU6cBS3","id":67178,"in_production":false,"languages":["en"],"last_air_date":"2019-01-18","last_episode_to_air":{"id":1675739,"name":"El torbellino","overview":"Frank, Pilgrim, Madani, Russo, Amy, Curt y la familia Schultz. La conclusión explosiva de la segunda temporada no escatima en munición.","vote_average":8.2,"vote_count":57,"air_date":"2019-01-18","episode_number":13,"episode_type":"finale","production_code":"","runtime":57,"season_number":2,"show_id":67178,"still_path":"/oqu7NwokmhEQRCplas6wixHJ9H2.jpg"},"name":"Marvel - The Punisher","next_episode_to_air":null,"networks":[{"id":213,"logo_path":"/wwemzKWzjKYJFfCeiB57q3r4Bcm.png","name":"Netflix","origin_country":""}],"number_of_episodes":26,"number_of_seasons":2,"origin_country":["US"],"original_language":"en","original_name":"Marvel's The Punisher","overview":"Un antiguo marine decidido a castigar a los asesinos de su familia termina atrapado en una conspiración militar.","popularity":14.5821,"poster_path":"/8OdIbo68mFE4w6jXA20aL8ocPsE.jpg","production_companies":[{"id":38679,"logo_path":"/v2y3LuLxYtW36hvLa8IDGQk3Oql.png","name":"Marvel Television","origin_country":"US"},{"id":19366,"logo_path":"/vOH8dyQhLK01pg5fYkgiS31jlFm.png","name":"ABC Studios","origin_country":"US"},{"id":156964,"logo_path":null,"name":"Bohemian Risk Productions","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"seasons":[{"air_date":"2017-11-17","episode_count":13,"id":78693,"name":"Temporada 1","overview":"Tras vengarse de los responsables de la muerte de su mujer y sus hijos, tomándose la justicia por su mano, Frank Castle descubre una conspiración que va mucho más allá del submundo criminal de Nueva York. Ahora conocido en toda la ciudad como el Castigador, debe descubrir la verdad sobre injusticias que afectan más familias aparte de la suya propia.","poster_path":"/ngg9srI1F4op7bRHX33QJhls9zq.jpg","season_number":1,"vote_average":8.0},{"air_date":"2019-01-18","episode_count":13,"id":115281,"name":"Temporada 2","overview":"Frank Castle, ex marine convertido en vigilante, llevaba una vida tranquila en la carretera hasta que, de repente, tiene que salvar a una joven chica de ser asesinada. A medida que se ve envuelto en el misterio que rodea a la joven y a aquellos que persiguen la información que ella posee, Castle atrae un nuevo objetivo a su espalda mientras nuevos y viejos enemigos le obligan a enfrentarse a si debe aceptar su destino y abrazar la vida como \"El Castigador\".","poster_path":"/kMYg7RbRYjzv9xQbkSg70G7fIM3.jpg","season_number":2,"vote_average":8.0}],"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Canceled","tagline":"","type":"Scripted","vote_average":8.127,"vote_count":2847}