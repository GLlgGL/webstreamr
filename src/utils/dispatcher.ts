import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
