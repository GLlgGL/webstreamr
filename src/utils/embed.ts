import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
