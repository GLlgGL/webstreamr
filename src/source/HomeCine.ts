import * as cheerio from 'cheerio';
import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getTmdbId, getTmdbNameAndYear, Id, TmdbId } from '../utils';
import { Source, SourceResult } from './Source';

export class HomeCine extends Source {
  public readonly id = 'homecine';

  public readonly label = 'HomeCine';

  public readonly contentTypes: ContentType[] = ['movie', 'series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.es, CountryCode.mx];

  public readonly baseUrl = 'https://www3.homecine.to';

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    super();

    this.fetcher = fetcher;
  }

  public async handleInternal(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const [name, year] = await getTmdbNameAndYear(ctx, this.fetcher, tmdbId, 'es');

    const pageUrl = await this.fetchPageUrl(ctx, name, year, tmdbId);
    if (!pageUrl) {
      return [];
    }

    const pageHtml = await this.fetcher.text(ctx, pageUrl);

    let linksHtml = pageHtml;
    if (tmdbId.season) {
      const episodeUrl = await this.fetchEpisodeUrl(pageHtml, tmdbId);
      if (!episodeUrl) {
        return [];
      }

      linksHtml = await this.fetcher.text(ctx, episodeUrl);
    }

    const title = tmdbId.season ? `${name} ${tmdbId.season}x${tmdbId.episode}` : `${name} (${year})`;

    const $ = cheerio.load(linksHtml);

    return $('.les-content a')
      .map((_i, el) => {
        let countryCode: CountryCode;
        if ($(el).text().toLowerCase().includes('latino')) {
          countryCode = CountryCode.mx;
        } else if ($(el).text().toLowerCase().includes('castellano')) {
          countryCode = CountryCode.es;
        } else {
          return [];
        }

        return { countryCode, title, url: new URL($('iframe', $(el).attr('href')).attr('src') as string) };
      }).toArray();
  };

  private readonly fetchPageUrl = async (ctx: Context, name: string, year: number, tmdbId: TmdbId): Promise<URL | undefined> => {
    const searchUrl = new URL(`/?s=${encodeURIComponent(name)}`, this.baseUrl);

    const html = await this.fetcher.text(ctx, searchUrl);

    const $ = cheerio.load(html);

    const keywords = [
      name,
      name.replace('-', 'â€“'),
    ];

    const urls: URL[] = [];
    keywords.map((keyword) => {
      urls.push(
        ...$(`a[oldtitle="${keyword} (${year})"], a[oldtitle="${keyword}"]`)
          .filter((_i, el) => $(el).siblings('#hidden_tip').find(`a[href$="release-year/${year}"]`).length !== 0)
          .map((_i, el) => new URL($(el).attr('href') as string))
          .toArray()
          .filter(url => tmdbId.season ? url.href.includes('/series/') : !url.href.includes('/series/')),
      );
    });

    return urls[0];
  };

  private readonly fetchEpisodeUrl = async (pageHtml: string, tmdbId: TmdbId): Promise<URL | undefined> => {
    const $ = cheerio.load(pageHtml);

    const urls = $('#seasons a')
      .map((_i, el) => new URL($(el).attr('href') as string))
      .toArray()
      .filter(url => url.href.endsWith(`-temporada-${tmdbId.season}-capitulo-${tmdbId.episode}`));

    return urls[0];
  };
}
