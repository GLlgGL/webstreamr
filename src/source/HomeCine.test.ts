import { createTestContext } from '../test';
import { FetcherMock, TmdbId } from '../utils';
import { HomeCine } from './HomeCine';

const ctx = createTestContext({ es: 'on', mx: 'on' });

describe('HomeCine', () => {
  let source: HomeCine;

  beforeEach(() => {
    source = new HomeCine(new FetcherMock(`${__dirname}/__fixtures__/HomeCine`));
  });

  test('handles non-uploaded movie gracefully', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(3176, undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handles non-existent episodes gracefully', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(1402, 1, 99));
    expect(streams).toHaveLength(0);
  });

  test('handle walking dead s1e1', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(1402, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('handle el camino', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(559969, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('handle marvel - the punisher', async () => {
    const streams = await source.handle(createTestContext({ es: 'on' }), 'movie', new TmdbId(67178, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('does not return mx content for es', async () => {
    const streams = await source.handle(createTestContext({ es: 'on' }), 'movie', new TmdbId(559969, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('does not return es content for mx', async () => {
    const streams = await source.handle(createTestContext({ mx: 'on' }), 'movie', new TmdbId(559969, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('handle here today', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(711475, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
