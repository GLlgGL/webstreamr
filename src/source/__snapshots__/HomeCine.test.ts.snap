// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HomeCine does not return es content for mx 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "mx",
      ],
      "title": "El Camino: Una película de Breaking Bad (2019)",
    },
    "url": "https://fastream.to/embed-kc5dvcy34x8p.html",
  },
]
`;

exports[`HomeCine does not return mx content for es 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "es",
      ],
      "title": "El Camino: Una película de Breaking Bad (2019)",
    },
    "url": "https://fastream.to/embed-c9tmfzsbfj0z.html",
  },
]
`;

exports[`HomeCine handle el camino 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "mx",
      ],
      "title": "El Camino: Una película de Breaking Bad (2019)",
    },
    "url": "https://fastream.to/embed-kc5dvcy34x8p.html",
  },
  {
    "meta": {
      "countryCodes": [
        "es",
      ],
      "title": "El Camino: Una película de Breaking Bad (2019)",
    },
    "url": "https://fastream.to/embed-c9tmfzsbfj0z.html",
  },
]
`;

exports[`HomeCine handle marvel - the punisher 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "es",
      ],
      "title": "Marvel - The Punisher 1x1",
    },
    "url": "https://fastream.to/embed-ujq9uojoftog.html",
  },
]
`;

exports[`HomeCine handle walking dead s1e1 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "mx",
      ],
      "title": "The Walking Dead 1x1",
    },
    "url": "https://fastream.to/embed-nkztf1memuca.html",
  },
  {
    "meta": {
      "countryCodes": [
        "es",
      ],
      "title": "The Walking Dead 1x1",
    },
    "url": "https://fastream.to/embed-5bv96pdifz0b.html",
  },
]
`;
